envoy/network/dns_resolver.h

```diff
@@ -26,7 +26,30 @@ class DnsResolverFactory : public Config::TypedFactory {
       const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) const PURE;
 
   std::string category() const override { return std::string(DnsResolverCategory); }
+
+  /**
+   * Initialize the related data for this type of DNS resolver.
+   * For some DNS resolvers, like c-ares, there are some specific data structure
+   * needs to be initialized before using it to resolve target.
+   */
+  virtual void initialize() {}
+
+  /**
+   * Cleanup the related data for this type of DNS resolver.
+   * For some DNS resolvers, like c-ares, there are some specific data structure
+   * needs to be cleaned up before terminates Envoy.
+   */
+  virtual void terminate() {}
 };
 
+/**
+ * Terminate the DNS resolver factories.
+ */
+static inline void terminateDnsResolverFactories() {
+  auto& factories = Registry::FactoryRegistry<Network::DnsResolverFactory>::factories();
```

Comment:
suggest making this a static method of DnsResolverFactory and put the impl in dns_resolver_util.cc, rather than putting an inline impl in the interface.